safekeeper/src/http/routes.rs

```diff
@@ -373,8 +378,53 @@ async fn dump_debug_handler(mut request: Request<Body>) -> Result<Response<Body>
         .await
```

Comment:
`resp` here will contain thousnads of "heavy" objects which can eat a lot of RAM resources. If you look at the `Response` struct in `debug_dump.rs`, you'll see this:

```
/// Response for debug dump request.
#[derive(Debug, Serialize, Deserialize)]
pub struct Response {
    pub start_time: DateTime<Utc>,
    pub finish_time: DateTime<Utc>,
    pub timelines: Vec<Timeline>, // <--- this Vec can contain 500MB+ of data
    pub timelines_count: usize,
    pub config: Config,
}
```

My idea is that we can reimplement this in a "lazy" way, and not prepare data until serialization. It should be possible to do something like this:

```
/// Response for debug dump request.
#[derive(Debug, Serialize, Deserialize)]
pub struct Response {
    pub start_time: DateTime<Utc>,
    pub finish_time: DateTime<Utc>,
    pub timelines: Vec<TimelineDumpSer>, // this will only contain Arc<Timeline>, which is lightweight
    pub timelines_count: usize,
    pub config: Config,
}

struct TimelineDumpSer {
    timeline: Arc<crate::timeline::Timeline>,
}

impl Serialize for TimelineDumpSer {
   fn serialize(&self) {
       let dump = build_from_tli(self.timeline);
       serde_json::serialize(dump)
   }
}
```

Reply:
Um, I have one question. @petuhovskiy 
I think the `Timeline` in the `Response` struct is NOT the `safekeeper::timeline::Timeline`, BUT the `safekeeper::debug_dump::Timeline` struct. https://github.com/neondatabase/neon/blob/main/safekeeper/src/debug_dump.rs#L79
Do you mean that this struct takes so much space on RAM?

Reply:
Yes, now we compute and store all `safekeeper::debug_dump::Timeline` structs in memory when constructing a `Response`, but it should be better to build `safekeeper::debug_dump::Timeline` from `safekeeper::timeline::Timeline` at the time of serialization.